jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java

```diff
@@ -305,35 +313,47 @@ public void putLongInObjectVolatile(Object obj, long offset, long value) {
 	}
 
 	public void putObjectInObject(Object obj, long offset, Object value) {
+		/*[IF JAVA_SPEC_VERSION >= 23]*/
+		unsafe.putReference(obj, offset, value);
+		/*[ELSE] JAVA_SPEC_VERSION >= 23 */
 		unsafe.putObject(obj, offset, value);
+		/*[ENDIF] JAVA_SPEC_VERSION >= 23 */
 	}
 
 	public void putObjectInObjectVolatile(Object obj, long offset, Object value) {
+		/*[IF JAVA_SPEC_VERSION >= 23]*/
+		unsafe.putReferenceVolatile(obj, offset, value);
+		/*[ELSE] JAVA_SPEC_VERSION >= 23 */
 		unsafe.putObjectVolatile(obj, offset, value);
+		/*[ENDIF] JAVA_SPEC_VERSION >= 23 */
 	}
 
 	public boolean compareAndSwapIntInObject(Object obj, long offset, int expected, int value) {
-/*[IF JAVA_SPEC_VERSION >= 9]*/
+		/*[IF JAVA_SPEC_VERSION >= 9]*/
 		return unsafe.compareAndSetInt(obj, offset, expected, value);
-/*[ELSE] JAVA_SPEC_VERSION >= 9 */
+		/*[ELSE] JAVA_SPEC_VERSION >= 9 */
 		return unsafe.compareAndSwapInt(obj, offset, expected, value);
-/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
+		/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
 	}
 
 	public boolean compareAndSwapLongInObject(Object obj, long offset, long expected, long value) {
-/*[IF JAVA_SPEC_VERSION >= 9]*/
+		/*[IF JAVA_SPEC_VERSION >= 9]*/
 		return unsafe.compareAndSetLong(obj, offset, expected, value);
-/*[ELSE] JAVA_SPEC_VERSION >= 9 */
+		/*[ELSE] JAVA_SPEC_VERSION >= 9 */
 		return unsafe.compareAndSwapLong(obj, offset, expected, value);
-/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
+		/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
 	}
 
 	public boolean compareAndSwapObjectInObject(Object obj, long offset, Object expected, Object value) {
-/*[IF JAVA_SPEC_VERSION >= 9]*/
+		/*[IF JAVA_SPEC_VERSION >= 9]*/
+		/*[IF JAVA_SPEC_VERSION >= 23]*/
+		return unsafe.compareAndSetReference(obj, offset, expected, value);
+		/*[ELSE] JAVA_SPEC_VERSION >= 23 */
 		return unsafe.compareAndSetObject(obj, offset, expected, value);
-/*[ELSE] JAVA_SPEC_VERSION >= 9 */
+		/*[ENDIF] JAVA_SPEC_VERSION >= 23 */
+		/*[ELSE] JAVA_SPEC_VERSION >= 9 */
 		return unsafe.compareAndSwapObject(obj, offset, expected, value);
-/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
+		/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
```

Comment:
Suggest `ELSEIF` instead of nesting:
```java
		/*[IF JAVA_SPEC_VERSION >= 23]*/
		return unsafe.compareAndSetReference(obj, offset, expected, value);
		/*[ELSEIF JAVA_SPEC_VERSION >= 9]*/
		return unsafe.compareAndSetObject(obj, offset, expected, value);
		/*[ELSE] JAVA_SPEC_VERSION >= 9 */
		return unsafe.compareAndSwapObject(obj, offset, expected, value);
		/*[ENDIF] JAVA_SPEC_VERSION >= 23 */
```